### PR TITLE
Fix codable configuration optionals decoding

### PR DESCRIPTION
**Pull request by TomaszLizer**
This PR handles issue with Optional types decoding using `CodableConfiguration`.
- Adds tests for validation of decoding behaviour of optional type
- Fixes edge case for `null` value for a key

## Context

Expected behaviour based on other examples - like Decodable types decoding is to treat missing key or `null` value as correct and decoding to nil.

Current implementation of `CodableConfiguration` treats `null` value for key as decoding error.

## Note
I was also considering fixing issue by minimally altering original extension by adding ` try !self.decodeNil(forKey: key)` to check however approach in PR seems cleaner to me.
Other considered fix:
```swift
@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
public extension KeyedDecodingContainer {
    func decode<T, C>(_: CodableConfiguration<T?, C>.Type, forKey key: Self.Key) throws -> CodableConfiguration<T?, C> {
        if self.contains(key), try !self.decodeNil(forKey: key) {
            let wrapper = try self.decode(CodableConfiguration<T, C>.self, forKey: key)
            return CodableConfiguration<T?, C>(wrappedValue: wrapper.wrappedValue)
        } else {
            return CodableConfiguration<T?, C>(wrappedValue: nil)
        }
    }
}

```

Closes Issue #1630

### CODE REVIEW

**Review by jmschonfeld — DISMISSED** [review]

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Comment by TomaszLizer** [comment]
@jmschonfeld Sure - will do that today :)  
Sorry for missing that, but I was unable to build in foundation framework mode at that time hence I guess that is the reason for that.
Will update if any issues faced.

**[EDIT]**: Updated - should be fine now. 
QQ however - is there I way I could simply build and test [`FOUNDATION_FRAMEWORK`](https://github.com/swiftlang/swift-foundation/blob/main/Foundation_Build_Process.md#foundation_framework-build) mode? I experimented a bit with adding flags during initial implementation, but only encountered failures. Actually at that time I was only able to build via `swift test` 🤔

**Comment by TomaszLizer** [comment]
@swift-ci please test

**[EDIT]:** Expected it won't work but had to try :D

File: Sources/FoundationEssentials/CodableWithConfiguration.swift [hunk 1]
```diff
@@ -53,14 +53,9 @@ public extension KeyedEncodingContainer {
...
 public extension KeyedDecodingContainer {
     func decode<T, C>(_: CodableConfiguration<T?, C>.Type, forKey key: Self.Key) throws -> CodableConfiguration<T?, C> {
-        if self.contains(key) {
-            let wrapper = try self.decode(CodableConfiguration<T, C>.self, forKey: key)
-            return CodableConfiguration<T?, C>(wrappedValue: wrapper.wrappedValue)
-        } else {
-            return CodableConfiguration<T?, C>(wrappedValue: nil)
-        }
+        let wrapper = try self.decodeIfPresent(CodableConfiguration<T?,C>.self, forKey: key)
+        return CodableConfiguration<T?, C>(wrappedValue: wrapper?.wrappedValue)
```
Comment (jmschonfeld) [inline Sources/FoundationEssentials/CodableWithConfiguration.swift:57]: Sorry if I missed this before, but now isn't `wrapper` a `CodableConfiguration<T?, C>`? If so, is there any reason not to just return that instead of creating a new `CodableConfiguration` from the wrapped value?

Comment (TomaszLizer) [reply Sources/FoundationEssentials/CodableWithConfiguration.swift:57]: Nah - type of `wrapper` is `Optional<CodableConfiguration<T?, C>>` hence it needs to be wrapped some more :D

**Comment by jmschonfeld** [comment]
> QQ however - is there I way I could simply build and test [FOUNDATION_FRAMEWORK](https://github.com/swiftlang/swift-foundation/blob/main/Foundation_Build_Process.md#foundation_framework-build) mode? I experimented a bit with adding flags during initial implementation, but only encountered failures. Actually at that time I was only able to build via swift test 🤔

At this time not currently, the `FOUNDATION_FRAMEWORK` build is a configuration used when building the full `Foundation.framework` and will fail to build without other pieces there. In general, building without that is sufficient (and is the only configuration needed on platforms like Linux + Windows) but every once in a while we just need a small tweak to something like this to get it to work. I'll make sure that's all sorted out, thanks for updating!

File: Sources/FoundationEssentials/CodableWithConfiguration.swift (hunk 1 above)
Comment (jmschonfeld) [reply Sources/FoundationEssentials/CodableWithConfiguration.swift:57]: Ah right yep that makes sense, thanks for explaining 🙂

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Review by jmschonfeld — APPROVED** [review]
